pkg/kvstore/allocator/allocator_test.go

```diff
@@ -0,0 +1,323 @@
+
+	saved := allocator.backoffTemplate.Factor
+	allocator.backoffTemplate.Factor = 1.0
+
+	// we should be out of id space here
+	_, new, err := allocator.Allocate(TestType(fmt.Sprintf("key%04d", maxID+1)))
+	c.Assert(err, Not(IsNil))
+	c.Assert(new, Equals, false)
+
+	allocator.backoffTemplate.Factor = saved
+
+	// allocate all IDs again using the same set of keys, refcnt should go to 2
+	for i := ID(1); i <= maxID; i++ {
+		key := TestType(fmt.Sprintf("key%04d", i))
+		id, new, err := allocator.Allocate(key)
+		c.Assert(err, IsNil)
+		c.Assert(id, Not(Equals), 0)
+		c.Assert(new, Equals, false)
+
+		// refcnt must now be 2
+		c.Assert(allocator.localKeys.keys[key.GetKey()].refcnt, Equals, uint64(2))
+	}
+
+	// Create a 2nd allocator, refill it
+	allocator2, err := NewAllocator(allocatorName, TestType(""), WithMax(maxID), WithSuffix("b"))
+	c.Assert(err, IsNil)
+	c.Assert(allocator2, Not(IsNil))
+
+	// allocate all IDs again using the same set of keys, refcnt should go to 2
+	for i := ID(1); i <= maxID; i++ {
+		key := TestType(fmt.Sprintf("key%04d", i))
+		id, new, err := allocator2.Allocate(key)
+		c.Assert(err, IsNil)
+		c.Assert(id, Not(Equals), 0)
+		c.Assert(new, Equals, false)
+
+		localKey := allocator2.localKeys.keys[key.GetKey()]
+		c.Assert(localKey, Not(IsNil))
+
+		// refcnt in the 2nd allocator is 1
+		c.Assert(localKey.refcnt, Equals, uint64(1))
+
+		allocator2.Release(key)
+	}
+
+	// release 2nd reference of all IDs
+	for i := ID(1); i <= maxID; i++ {
+		allocator.Release(TestType(fmt.Sprintf("key%04d", i)))
+	}
+
+	// refcnt should be back to 1
+	for i := ID(1); i <= maxID; i++ {
+		key := TestType(fmt.Sprintf("key%04d", i))
+		c.Assert(allocator.localKeys.keys[key.GetKey()].refcnt, Equals, uint64(1))
+	}
+
+	// running the GC should not evict any entries
+	allocator.runGC()
+
+	v, err := kvstore.ListPrefix(allocator.idPrefix)
+	c.Assert(err, IsNil)
+	c.Assert(len(v), Equals, int(maxID))
+
+	// release final reference of all IDs
+	for i := ID(1); i <= maxID; i++ {
+		allocator.Release(TestType(fmt.Sprintf("key%04d", i)))
+	}
+
+	// running the GC should evict all entries
+	allocator.runGC()
+
+	v, err = kvstore.ListPrefix(allocator.idPrefix)
+	c.Assert(err, IsNil)
+	c.Assert(len(v), Equals, 0)
+
+	allocator.DeleteAllKeys()
+	allocator.Delete()
+	allocator2.Delete()
+}
+
+func (s *AllocatorSuite) TestAllocateCached(c *C) {
+	testAllocator(c, ID(256), randStringRunes(12), "a") // enable use of local cache
+}
+
+func (s *AllocatorSuite) TestKeyToID(c *C) {
+	allocatorName := randStringRunes(12)
+	a, err := NewAllocator(allocatorName, TestType(""), WithSuffix("a"))
+	c.Assert(err, IsNil)
+	c.Assert(a, Not(IsNil))
+
+	c.Assert(a.keyToID(path.Join(allocatorName, "invalid"), false), Equals, NoID)
+	c.Assert(a.keyToID(path.Join(a.idPrefix, "invalid"), false), Equals, NoID)
+	c.Assert(a.keyToID(path.Join(a.idPrefix, "10"), false), Equals, ID(10))
+}
+
+//func testParallelAllocator(c *C, maxID ID, allocatorName string, suffix string) {
```

Comment:
There's still some commented-out tests here.

Reply:
Yes, these tests are failing in ginkgo CI. I can run them for hours locally. I haven't figured out what exactly is different when ran on Jenkins. I'll add a comment explaining why they are commented out.